Proj/C++/Proj2/DoxygenProject2/texasHoldemVer7/man/man3/Dealer.3
.TH "Dealer" 3 "Fri Jun 7 2019" "texasHoldem" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Dealer
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Dealer\&.h>\fP
.SS "Classes"

.in +1c
.ti -1c
.RI "class \fBNegativeNumber\fP"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBDealer\fP ()"
.br
.ti -1c
.RI "\fB~Dealer\fP ()"
.br
.ti -1c
.RI "\fBDealer\fP (int n)"
.br
.ti -1c
.RI "void \fBshuffle\fP ()"
.br
.ti -1c
.RI "void \fBsetPlayers\fP (int n)"
.br
.ti -1c
.RI "void \fBsetCards\fP (\fBCard\fP *c)"
.br
.ti -1c
.RI "void \fBsetMyCards\fP (\fBCard\fP *mc)"
.br
.ti -1c
.RI "void \fBsetRound\fP (int r)"
.br
.ti -1c
.RI "void \fBinitiateRound\fP ()"
.br
.ti -1c
.RI "int \fBdecideWinner\fP ()"
.br
.ti -1c
.RI "void \fBcompleteRound\fP ()"
.br
.ti -1c
.RI "void \fBdealPreflop\fP ()"
.br
.ti -1c
.RI "void \fBdealFlop\fP ()"
.br
.ti -1c
.RI "void \fBdealTurn\fP ()"
.br
.ti -1c
.RI "void \fBdealRiver\fP ()"
.br
.ti -1c
.RI "int \fBgetPotAmount\fP () const"
.br
.ti -1c
.RI "int \fBgetInitContAmount\fP () const"
.br
.ti -1c
.RI "int \fBgetRound\fP () const"
.br
.ti -1c
.RI "\fBPlayer\fP * \fBgetPlayers\fP () const"
.br
.ti -1c
.RI "\fBCard\fP * \fBgetCards\fP () const"
.br
.ti -1c
.RI "int \fBgetNumPlayers\fP () const"
.br
.ti -1c
.RI "void \fBdisplayCards\fP (\fBCard\fP *cards)"
.br
.ti -1c
.RI "void \fBdisplay\fP ()"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 18 of file Dealer\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Dealer::Dealer ()"

.PP
Definition at line 20 of file Dealer\&.cpp\&.
.SS "Dealer::~Dealer ()\fC [inline]\fP"

.PP
Definition at line 33 of file Dealer\&.h\&.
.SS "Dealer::Dealer (int n)"

.PP
Definition at line 26 of file Dealer\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void Dealer::completeRound ()"

.PP
Definition at line 250 of file Dealer\&.cpp\&.
.SS "void Dealer::dealFlop ()"

.PP
Definition at line 147 of file Dealer\&.cpp\&.
.SS "void Dealer::dealPreflop ()"

.PP
Definition at line 105 of file Dealer\&.cpp\&.
.SS "void Dealer::dealRiver ()"

.PP
Definition at line 189 of file Dealer\&.cpp\&.
.SS "void Dealer::dealTurn ()"

.PP
Definition at line 180 of file Dealer\&.cpp\&.
.SS "int Dealer::decideWinner ()"

.PP
Definition at line 201 of file Dealer\&.cpp\&.
.SS "void Dealer::display ()"

.PP
Definition at line 275 of file Dealer\&.cpp\&.
.SS "void Dealer::displayCards (\fBCard\fP * cards)"

.PP
Definition at line 258 of file Dealer\&.cpp\&.
.SS "\fBCard\fP* Dealer::getCards () const\fC [inline]\fP"

.PP
Definition at line 51 of file Dealer\&.h\&.
.SS "int Dealer::getInitContAmount () const\fC [inline]\fP"

.PP
Definition at line 48 of file Dealer\&.h\&.
.SS "int Dealer::getNumPlayers () const\fC [inline]\fP"

.PP
Definition at line 52 of file Dealer\&.h\&.
.SS "\fBPlayer\fP* Dealer::getPlayers () const\fC [inline]\fP"

.PP
Definition at line 50 of file Dealer\&.h\&.
.SS "int Dealer::getPotAmount () const\fC [inline]\fP"

.PP
Definition at line 47 of file Dealer\&.h\&.
.SS "int Dealer::getRound () const\fC [inline]\fP"

.PP
Definition at line 49 of file Dealer\&.h\&.
.SS "void Dealer::initiateRound ()"

.PP
Definition at line 89 of file Dealer\&.cpp\&.
.SS "void Dealer::setCards (\fBCard\fP * c)\fC [inline]\fP"

.PP
Definition at line 37 of file Dealer\&.h\&.
.SS "void Dealer::setMyCards (\fBCard\fP * mc)\fC [inline]\fP"

.PP
Definition at line 38 of file Dealer\&.h\&.
.SS "void Dealer::setPlayers (int n)"

.PP
Definition at line 78 of file Dealer\&.cpp\&.
.SS "void Dealer::setRound (int r)\fC [inline]\fP"

.PP
Definition at line 39 of file Dealer\&.h\&.
.SS "void Dealer::shuffle ()"

.PP
Definition at line 56 of file Dealer\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for texasHoldem from the source code\&.


Proj/C++/Proj2/texasHoldem_Ver4/Dealer.cpp
/* 
 * File:   Dealer.cpp
 * Author: Byoung Mo Lee
 * Created on May 17, 2019 00:19 AM
 * Purpose: Texas Holdem
 * 
 */
#include <string>
#include <iostream>
#include <cstdlib>
#include <memory>
#include "Dealer.h"



using namespace std;

Dealer::Dealer(int n) {
    round=0;
    potAmount=0;
    initialContributeAmount=50;
    numPlayers=n;
    setPlayers(n);
    DeckOfCards();
}

void Dealer::DeckOfCards() {
    cards=new Card [52];
    for(int j=0;j<4;j++){
        for(int i=0;i<13;i++){
            cards[13*j+i].setCard(i+1,j);
        }
    }
    for(int j=0;j<4;j++){
        for(int i=0;i<13;i++){
            if(cards[13*j+i].getFace()==1) cards[13*j+i].setFace(14);
            //cards[13*j+i].displayCard(); //test for deckOfCards
        }
        //cout << endl;
    }
}



void Dealer::shuffle(){
    // Initialize seed randomly 
    srand(time(0)); 
  
    for (int i=0; i<52 ;i++) 
    { 
        // Random for remaining positions. 
        int r = i + (rand() % (52 -i)); 
  
        swap(cards[i], cards[r]);
     }
    
    //for(int i=0;i<52;i++){
    //    cards[i].displayCard();
    //} 
    //cout << endl;// test for shuffle function
}

void Dealer::setPlayers(int n){
    players=new Player[n];
    numPlayers=n;
    for(int i=0;i<n;i++){
        players[i].setOrder(i);
    }
}

void Dealer::initiateRound(){
    //int bigBlind=initialContributeAmount*2;
    //int smallBlind=initialContributeAmount;
    int scan=0;
    int found=0;
    
    while(scan<numPlayers && !found){
        if(players[scan].getOrder()==0) found=1;
        else scan++;
    }
    
    players[scan].putMoneyToPot(initialContributeAmount*2);
    potAmount+=initialContributeAmount*2;
    players[scan+1].putMoneyToPot(initialContributeAmount);
    potAmount+=initialContributeAmount;
    
}//Inheritance, contribute money from big blind and small blind

void Dealer::dealPreflop(){
    int scan=0;
    int dealOrder;
    bool found=0;
    
    while(!found && scan<7){
        if(players[scan].getOrder()==0) {
            //cout << players[count].getOrder() << endl; //test
            found=1;
        }
        else scan++;
    }
    
    if(!found) cout << "Error: cannot find the player's order=0" << endl;
    
    for(int i=0;i<2*numPlayers;i++){   
        dealOrder=(scan+i)%numPlayers;  //we distribute the cards to big blind first
        players[dealOrder].addCards(cards+i,1);
    }
    
    for(int i=0;i<numPlayers;i++){
        displayCards(players[i].getMycards(),2);
        cout << "            " ;
    }
    cout << endl << endl;
    
    for(int i=0;i<numPlayers-2;i++){
        cout << "check(1) or die(0): " << endl;
        players[scan+i].setHands(players[scan+i].getMycards());
        if(players[scan+i].getPoints()>0) {
            players[scan+1+i].putMoneyToPot(initialContributeAmount);
            potAmount+=initialContributeAmount;
        }
        else players[scan+i].setStatus(0);
    }
}
   
    
void Dealer::dealFlop(){
    int n=numPlayers*2;
    int bet=50;
    int found=0;
    int scan=0;
    
    cout << "           ";
    displayCards(cards+n+1,3);
    cout << endl;
    
    while(!found && scan<7){
        if(players[scan].getOrder()==0) {
            //cout << players[count].getOrder() << endl; //test
            found=1;
        }
        else scan++;
    }
    
    for(int i=0;i<numPlayers;i++){
        if(players[scan+i].getStatus()) {
            players[scan+i].addCards(cards+n+1,3);  //Since before the flop dealer burn a card 
            cout << "check(1) or die(0)" << endl;
            if(players[scan+i].getPoints()>0){
                cout << "check(call)" << endl;
                players[scan+i].putMoneyToPot(bet);
                potAmount+=bet;
            }
            else cout << "Player" << scan+i << " died" << endl;
        }
    }
}

void Dealer::dealTurn() {
    int n=numPlayers*2+4;                  //what are turned
    
    for(int i=0;i<numPlayers;i++){   
        players[i].addCards(cards+n+1,1);  //Since before the turn dealer burn a card 
    }
    displayCards(cards+n+1,1);
}
void Dealer::dealRiver() {
    int n=numPlayers*2+4+2;                 //what are turned
    
    for(int i=0;i<numPlayers;i++){   
        players[i].addCards(cards+n+1,1);  //Since before the river dealer burn a card 
    }
    displayCards(cards+n+1,1);
}



void Dealer::decideWinner(){
    int* points=new int[numPlayers];
    for(int i=0;i<numPlayers;i++){
        players[i].getHands(players[i].getMycards(),players[i].getNumCards());
        cout << "hands[" << i << "]:" << players[i].getHands(players[i].getMycards(),players[i].getNumCards()) << endl;
        points[i]=players[i].getPoints();
        cout << "points[" << i << "]:" << players[i].getPoints() << endl;
    }
    selectionSort(points,0,7);
    cout << "test1" << endl;
    int count=0;                      //Since we need to split for tie
    int max=points[0];
    
    for(int i=0;i<numPlayers-1;i++){
        if(!(points[0]-points[i+1])) count++;
    }
    
    int* pos=new int [count];
    int scan=0;
    int index=0;
    cout << "test2" << endl;
    while(index<count && scan<numPlayers){
        if(players[scan].getPoints()==max) {
            pos[index]=scan;
            index++;
        }
        scan++;        
    }
    cout << "test3" << endl;
    for(int i=0;i<count;i++){
        players[pos[i]].takeMoneyFromPot(potAmount/count);
        cout << "player" << i << " won" << potAmount/count << endl;
    }
    
    delete[] points;
    delete[] pos;
}

void Dealer::completeRound(){
    for(int i=0;i<numPlayers;i++){
        players[i].setOrder((players[i].getOrder()+1)%numPlayers);
    }
    potAmount=0;
    round+=1;
} // rotate order by using Operator Overloading (++ operator go back to zero for the last player

void Dealer::displayCards(Card* cards,int n){
    for(int i=0;i<n;i++){
        cards[i].displayCard();
    }
} 

void Dealer::selectionSort(int* array,int begin, int size){
    int maxIndex, maxValue;
    int end=begin+size;
    for(int start=begin;start<end-1;start++){
        maxIndex=start;
        maxValue=array[start];
        for(int index=start+1;index<end;index++){
            if(array[index]>maxValue){
                maxValue=array[index];
                maxIndex=index;
            }
        }
        swap(array[maxIndex],array[start]);
        swap(array[maxIndex],array[start]);
    }
}
